File: lab7/ge.c
int ge_fw(float *matrix, int rows, int cols, float *matrix_out) {

  int r=0;
  int c=0;
  
  int i = 0;
  int z = 0;
  float scalar = 0;
  
  int rowcounter = 0;
  int columncounter = 0;
  float swaptemp = 0;
  int scalarcounter = 0;
  
  
  if (matrix==NULL) {
    return -1;
  }
  if (matrix_out==NULL) {
    return -1;
  }
  
  /*copy the matrix in to matrix out */
  for(r=0;r<rows;r=r+1) {
    for(c=0;c<cols;c=c+1) {
       matrix_out[r*cols + c] = matrix[r*cols + c];
    }
  }
  
  r = 0;
  c = 0;
  
  while ((c < cols) && (r < rows)){
   
    if (matrix_out[r*cols + c] == 0){
      
      /*iterate the rows for swaps*/
      for (i = 0; i<rows;i++){
      
         
        if (matrix_out[i*cols+c] != 0){
         /*swap*/
         swaptemp = matrix_out[r*cols +c];
         matrix_out[r*cols + c] = matrix_out[i*cols + c];
         matrix_out[i*cols + c] = swaptemp;
       }
       
      }
      
    }
    
    /*get zeroes underneath the first value in the column*/
    for (rowcounter = r + 1; rowcounter < rows; rowcounter++){
      
      scalarcounter = 0;
      
      
      while (scalarcounter < r+1){
        
        if (matrix_out[rowcounter*cols + c] != 0){
          
          /*keep the scalar multiple*/
          scalar = (matrix_out[rowcounter*cols + c] / matrix_out[r*cols + c]);
          
          /*apply the subtraction across all column*/
          for (columncounter=0; columncounter < cols; columncounter++){ 
          
            matrix_out[rowcounter*cols + columncounter] = matrix_out[rowcounter*cols + columncounter] - (scalar * matrix_out[r*cols + columncounter]);
    
          }
          
        }
        
        else{
          
          scalar = 0;
        }
        
        scalarcounter = scalarcounter + 1;
      }
     
    }
    
    r = r + 1;
    c = c + 1;
  } 
  return 0;
}



int ge_bw(float *matrix, int rows, int cols, float *matrix_out){
  int r = 0;
  int c = 0;
  int i = 0;
  int coeff = 0;
  int scalarcounter = 0;
  int columncounter = 0;
  int scalar = 0;
  
  if (matrix==NULL) {
    return -1;
  }
  if (matrix_out==NULL) {
    return -1;
  }
  
  /*copy the matrix in to matrix out*/
  for(r=0;r<rows;r=r+1) {
    for(c=0;c<cols;c=c+1) {
       matrix_out[r*cols + c] = matrix[r*cols + c];
    }
  }
  
  r = rows - 1;
  c = cols - 2;
  
  while (c >= 0 && r >= 0){
    
    /*normalize*/
    if (matrix_out[r*cols + c] != 1 && matrix_out[r*cols + c] != 0){
      coeff = matrix_out [r*cols + c];
      
      for (i = 0; i < cols; i++){
        if (matrix_out[r*cols + i] != 0){
          matrix_out[r*cols + i] = (matrix_out[r*cols + i] / coeff);
        }
      }
      
    }
    
    /*add multiples of last row */
    scalarcounter = r - 1;
    
    while (scalarcounter >=0){
      if (matrix_out[r*cols + c] != 0){
        scalar = matrix_out[scalarcounter*cols + c] / matrix_out[r*cols + c];
      }
      
      else{
        
        scalar = 0;
      }
      
      for (columncounter = 0; columncounter < cols; columncounter++){
        matrix_out[scalarcounter*cols + columncounter] = matrix_out[scalarcounter*cols + columncounter] - (scalar*matrix_out[r*cols + columncounter]);
        
      
      }
      scalarcounter = scalarcounter - 1;
    }
    
    r = r - 1;
    c = c - 1;
  }
  
  return 0;
  
}

File: lab7/ge.py
# coding: utf-8
def ge_fw(matrix):
  
  rows = len(matrix)
  columns = len(matrix[0])

 

  rowiterator = 0
  columniterator = 0
  
  while (columniterator < columns) and (rowiterator < rows):   
    
    #find the first non zero row
    if matrix [rowiterator][columniterator] == 0:
      for i in range(rowiterator, len(matrix)):
        if matrix[i][columniterator] != 0:
          #swap
          matrix[rowiterator], matrix[i] = matrix[i], matrix[rowiterator]
          break
      
    #get zeroes underneath the first value in the column
    for rowcounter in range(rowiterator + 1,rows):
      scalarcounter = 0
    

      while (scalarcounter < rowiterator+1):
        
        if matrix[rowcounter][columniterator] != 0:
          #keep the scalar multiple
          scalar = matrix[rowcounter][columniterator] / matrix[rowiterator][columniterator]
      
  
            
          #apply the subtraction across all columns 
          for columncounter in range(0, columns):
          
            matrix[rowcounter][columncounter] = matrix[rowcounter][columncounter] - (scalar * matrix[rowiterator][columncounter])
 
  
        scalarcounter = scalarcounter + 1

    
    rowiterator = rowiterator + 1
    columniterator = columniterator + 1 
    
  return matrix 
  
  
def ge_bw(matrix):
  rows = len(matrix)
  
  columns = len(matrix[0])
  
  #find the last non zero row at the bottom
  if matrix [rows-1][columns-1] == 0:
    for i in range(rows -1, -1, -1):
      if matrix[i][columns-1] != 0:
        lastrow = i
  else:
    lastrow = rows
  
        
  rowiterator = rows - 1
  columniterator = columns - 2 
  
  while (columniterator >= 0) and (rowiterator >= 0):

      
    #normalize, divide by coeff
    if matrix[rowiterator][columniterator] != 1 and matrix[rowiterator][columniterator] != 0:
      
      coeff = matrix [rowiterator][columniterator]
      
      for i in range (len(matrix[0])):
        if matrix[rowiterator][i] != 0:
          
          matrix[rowiterator][i] = matrix[rowiterator][i] / coeff
  


    #add multiples of last row
    
    scalarcounter = rowiterator - 1

    while (scalarcounter >= 0):
      
      if matrix[rowiterator][columniterator] != 0 :
      
        scalar = matrix[scalarcounter][columniterator] / matrix[rowiterator][columniterator]
      
      else:
        
        scalar = 0
      
    
      for columncounter in range(0, columns):
        
          matrix[scalarcounter][columncounter] = matrix[scalarcounter][columncounter] - (scalar * matrix[rowiterator][columncounter])       
          
      scalarcounter = scalarcounter - 1
    
    #print "subtract"
    #for i in range (rows):
      #print matrix[i]
        
    rowiterator = rowiterator - 1
    columniterator = columniterator - 1

  return matrix 
